### PR TITLE
Warn for missing returns with explicit Any return types

### LINKED ISSUE #16095: Add a NoImplicit return type specifier.
**Feature**

Example:
```
cdef foo(): # Type: Any
  ...
  return ...

def bar() -> Any:
   if bool:
      return foo()
   else:
     var = foo() + 1
     # forgot to return so implicitly return None which is a valid Any type
     
```
 mypy won't flag this as an error
 
 Proposed functionality example:
```
cdef foo(): # Type: Any
  ...
  return ...

def bar() -> NoImplicit[Any]: # the new function specific NoImplicit type specifier
   if bool:
      return foo()
   else:
     var = foo() + foo()
     # forgot to return
     
``` 
mypy flags it. error: Implicit return [None]

sometimes when using cython functions it is difficult to specify the return type and sometimes the exact type is not important. I just want to specify that the return is not implicit for only some functions in a large codebase.

Its possible to cast the return type and specify but it seems overly complex and messy when does this way:

```
cdef foo():
  ...
  return ...
  
def bar() -> List[Union[int, bytes, float]], List[Union[str, int, bytes]
   if bool:
     return cast(List[Union[int, bytes, float]], List[Union[str, int, bytes], foo())
   else:
     var = foo() + foo()
     # forgot to return. error: Missing return statement  [return]

```
And then imagine how cluttered it gets after calling foo hundreds of times across the codebase

### PR DESCRIPTION
**Pull request by hauntsaninja**
As discussed in https://github.com/python/mypy/issues/7511, mypy's `--warn-no-return` isn't really a question of type safety. It does however enforce a PEP 8 rule that is "dear to Guido's heart", and I think we should enforce it for functions that explicitly return Any as well.

Fixes #16095

### CODE REVIEW

**Comment by hauntsaninja** [comment]
Hmm, more hits than I was hoping for. Most of them are PEP 8 violations, the few that are not should straightforwardly be `-> None`. Still, might eat more into our disruption budget than we want.

File: test-data/unit/check-warnings.test [hunk 1]
```diff
@@ -140,6 +140,24 @@ def f() -> int:
...
+
+def maybe() -> bool: ...
+
+def implicit(x: Any):
+    if maybe():
+        return x
+
+def explicit(x: Any) -> Any:  # E: Missing return statement
+    if maybe():
+        return x
```
Comment (sobolevn) [inline test-data/unit/check-warnings.test:156]:
```suggestion
def explicit(x: Any) -> Any:  # E: Missing return statement
    if maybe():
        return x
        
def explicit_with_none(x: Any) -> Any:
    if maybe():
        return x
    return None
    
def explicit_with_just_return(x: Any) -> Any:
    if maybe():
        return x
    return
```

**Review by sobolevn — COMMENTED** [review]
Let's also test correct cases to be correct.

**Comment by jakethesnake420** [comment]
bump this

**Comment by hauntsaninja** [comment]
I think this is too disruptive for too little benefit

**Pull request closed without merging**